### PR TITLE
[Merged by Bors] - Run weak coin only for following vote rounds and after votes were received

### PR DESCRIPTION
**Pull request by dshulyak**
## Motivation

It bubbled up some time ago and also flagged by @countvonzero that there is a mistake in weak coin execution order. It needs to be executed *after* vote was sent to the network. Otherwise, a byzantine actor can negate the weak coin effect by adjusting its own vote.

Also, votes from the last round were ignored fully.

## Changes
- run weak coin after receiving votes for the current round
- change timeouts accordingly
- don't return errors from goroutines spawned in runConsensusPhase, as it will shutdown whole tortoise beacon
- lowercases some logs

## TODO

- [ ] Explain motivation or link existing issue(s)
- [ ] Test changes and document test plan
- [ ] Update documentation as needed

## DevOps Notes

- [x] This PR does not require configuration changes (e.g., environment variables, GitHub secrets, VM resources)
- [x] This PR does not affect public APIs
- [x] This PR does not rely on a new version of external services (PoET, elasticsearch, etc.)
- [x] This PR does not make changes to log messages (which monitoring infrastructure may rely on)

### CODE REVIEW

**Review by nkryuchkov — APPROVED** [review]

**Comment by dshulyak** [comment]
bors merge

**Comment by dshulyak** [comment]
weird, i thought that the beacon is disabled, but apparently it is not disabled for block builder
https://kibana-tests.spacemesh.io/goto/aaf4659c6bd032345de6002a0295393e

**Comment by dshulyak** [comment]
bors try

**Comment by dshulyak** [comment]
The test failed because node was expected to produce blocks, but because beacon wasn't found it skipped block production. There are other errors such as atx is not found but those are not the root cause of the failure and doesn't seem to be a new thing in general. Tortoise beacon probably wasn't disabled completely.

There are also a lot of `received malformed first voting message, already received a voting message for this pk and round` which is strange. This should be only if previously received messages were evicted from duplicates cache. But i don't see how that can be related to this change.

File: tortoisebeacon/tortoise_beacon.go [hunk 1]
```diff
@@ -503,84 +503,61 @@ func (tb *TortoiseBeacon) runConsensusPhase(ctx context.Context, epoch types.Epo
...
-
 	for round := firstRound; round <= tb.lastRound(); round++ {
-		// always use coinflip from the previous round for current round.
-		// round 1 is running without coinflip (e.g. value is false) intentionally
 		round := round
-		previousCoinFlip := coinFlip
+		votes := ownVotes
 		tb.eg.Go(func() error {
 			if round == firstRound {
 				if err := tb.sendProposalVote(ctx, epoch); err != nil {
```
Comment (countvonzero) [inline tortoisebeacon/tortoise_beacon.go:498]: sendProposalVote() should initialize ownVotes

File: tortoisebeacon/tortoise_beacon.go [hunk 2]
```diff
@@ -503,84 +503,61 @@ func (tb *TortoiseBeacon) runConsensusPhase(ctx context.Context, epoch types.Epo
...
-
-			return nil
-		})
-
-		tb.eg.Go(func() error {
-			tb.startWeakCoinRound(ctx, epoch, round)
 			return nil
 		})
+		if round != firstRound {
+			tb.eg.Go(func() error {
```
Comment (countvonzero) [inline tortoisebeacon/tortoise_beacon.go]: i think there is still a potential race condition here. a malicious peer can still wait till the weakcoin round to finish before it send out the votes, and their late votes will still be counted in this round.

i think we should
- send vote
- wait for the duration for vote to come in
- start weak coin (this is the cutoff time for counting votes in this round)
- count votes without weak coin
- finish weak coin
- count undecided votes with weak coin
- the late vote received after weak coin started will be added to the next round

File: tortoisebeacon/db_test.go [hunk 3]
```diff
@@ -0,0 +1,13 @@
+package tortoisebeacon
+
+import (
+	"testing"
+
+	"github.com/spacemeshos/go-spacemesh/database"
+	"github.com/spacemeshos/go-spacemesh/log/logtest"
+)
+
+func newMemDB(tb testing.TB) *DB {
```
Comment (countvonzero) [inline tortoisebeacon/db_test.go]: is this used?

Comment (dshulyak) [reply tortoisebeacon/db_test.go]: it was used in the test which i didn't submit in this pr. can remove

File: tortoisebeacon/tortoise_beacon.go (hunk 1 above)
Comment (dshulyak) [reply tortoisebeacon/tortoise_beacon.go:498]: why? in current code votes are initialized by calcVotes after received all first round messages. it is not correct?

Comment (dshulyak) [reply tortoisebeacon/tortoise_beacon.go:498]: i think sendProposalVote will just use initial votes for first voting message that were formed from all proposals

File: tortoisebeacon/tortoise_beacon.go [hunk 4]
```diff
@@ -503,84 +503,61 @@ func (tb *TortoiseBeacon) runConsensusPhase(ctx context.Context, epoch types.Epo
...
+				tb.startWeakCoinRound(ctx, epoch, round)
+				return nil
+			})
+		}
 
 		select {
 		case <-timer.C:
 			timer.Reset(tb.config.VotingRoundDuration + tb.config.WeakCoinRoundDuration)
```
Comment (countvonzero) [inline tortoisebeacon/tortoise_beacon.go]: this needs to adjust too.

are we sure in the test params that voting round duration is a lot longer than weakcoin duration?

File: tortoisebeacon/tortoise_beacon.go (hunk 2 above)
Comment (dshulyak) [reply tortoisebeacon/tortoise_beacon.go]: that makes sense

File: tortoisebeacon/tortoise_beacon.go (hunk 4 above)
Comment (dshulyak) [reply tortoisebeacon/tortoise_beacon.go]: > this needs to adjust too.

you mean if votes will have to be tallied twice?

> are we sure in the test params that voting round duration is a lot longer than weakcoin duration?

it is longer by 10-15s. why do you think that it should be *a lot* longer?

File: tortoisebeacon/tortoise_beacon.go (hunk 1 above)
Comment (countvonzero) [reply tortoisebeacon/tortoise_beacon.go:498]: can you point me to the code where this node's own proposal is added to votesMargin? i am lost in the sea of code now...

Comment (countvonzero) [reply tortoisebeacon/tortoise_beacon.go:498]: > why? in current code votes are initialized by calcVotes after received all first round messages. it is not correct?

it is, but that's only counting votes the node received. i cannot find where this node's own initial proposal/vote was added.

Comment (dshulyak) [reply tortoisebeacon/tortoise_beacon.go:498]: it just uses incoming proposals for payload https://github.com/spacemeshos/go-spacemesh/blob/782b11dcdd98ee10c4c43375f82a6236a192715c/tortoisebeacon/tortoise_beacon.go#L629-L636

Comment (dshulyak) [reply tortoisebeacon/tortoise_beacon.go:498]: https://github.com/spacemeshos/go-spacemesh/blob/782b11dcdd98ee10c4c43375f82a6236a192715c/tortoisebeacon/tortoise_beacon.go#L486-L489

**Comment by countvonzero** [comment]
> The test failed because node was expected to produce blocks, but because beacon wasn't found it skipped block production. There are other errors such as atx is not found but those are not the root cause of the failure and doesn't seem to be a new thing in general. Tortoise beacon probably wasn't disabled completely.

i wonder whether this is a timing/duration params issue.
this PR didn't seem to delay the time beacon is generated and just shifts the order of operations around inside the time frame.

**Comment by dshulyak** [comment]
> i wonder whether this is a timing/duration params issue.
this PR didn't seem to delay the time beacon is generated and just shifts the order of operations around inside the time frame.

yes very confusing. if anything it became a bit shorter since the whole tortoise beacon is shorter. previously it was, proposal + first + wc + n-1*(round + wc) and now first wc is gone.

**Comment by countvonzero** [comment]
> yes very confusing. if anything it became a bit shorter since the whole tortoise beacon is shorter. previously it was, proposal + first + wc + n-1*(round + wc) and now first wc is gone.

hmm.. i think it is
before: 
`FirstVotingRoundDuration+WeakCoinRoundDuration + (K-1)(VotingRoundDuration+WeakCoinRoundDuration)`

after: 
`FirstVotingRoundDuration + (K-1)(VotingRoundDuration+WeakCoinRoundDuration+timeSpentCalculatingVotes)`

unless `timeSpentCalculatingVotes` is significant, i agree it should be shorter

**Comment by countvonzero** [comment]
> unless timeSpentCalculatingVotes is significant, i agree it should be shorter

actually i am wrong. the (K-1) multiplier probably did make the beacon generation time longer.

File: tortoisebeacon/tortoise_beacon.go (hunk 4 above)
Comment (countvonzero) [reply tortoisebeacon/tortoise_beacon.go]: oh. i just mean that in this order of operation, VotingRoundDuration includes weakCoinRoundDuration. 
conceptually a voting round includes sending votes, generating weak coin, and counting votes. so i thought VotingRoundDuration covers all those.

File: tortoisebeacon/tortoise_beacon.go (hunk 1 above)
Comment (countvonzero) [reply tortoisebeacon/tortoise_beacon.go:498]: @dshulyak thanks. but this only serves to send our own proposal to peers, and in later encoding for following votes.
i still don't see where `tb.votesMargin` is updated with own initial votes.
`tb.votesMargin` seems to be only updated when it receives a gossip message from peers.

**Comment by dshulyak** [comment]
> timeSpentCalculatingVotes

it is very low, especially with the amount of nodes in tests. i have inprocess test with 50 nodes, it takes around ~1ms for calcVotes to complete.

File: tortoisebeacon/tortoise_beacon.go (hunk 1 above)
Comment (dshulyak) [reply tortoisebeacon/tortoise_beacon.go:498]: ah sorry i understood now what we were talking about. everything that is Broadcast'ed will be sent to the local handler for the protocol as well. So, what is sent in `sendProposalVote` will be also received HandleSerializedFirstVotingMessage. that's sort of a feature of p2p framework

**Comment by dshulyak** [comment]
bors try

File: tortoisebeacon/tortoise_beacon.go (hunk 1 above)
Comment (countvonzero) [reply tortoisebeacon/tortoise_beacon.go:498]: OMG... i'm learning something new about p2p every day. cool! thanks for your patience @dshulyak

Comment (countvonzero) [reply tortoisebeacon/tortoise_beacon.go:498]: but then, does that mean we add our own proposal to the valid list twice? once in `proposalPhaseImpl` and once in the proposal handler? @dshulyak

**Comment by dshulyak** [comment]
previous test failure has the same root cause as in https://github.com/spacemeshos/go-spacemesh/issues/2764 . definitely doesn't have anything to do with tortoise beacon.

bors try

File: tortoisebeacon/tortoise_beacon.go (hunk 1 above)
Comment (dshulyak) [reply tortoisebeacon/tortoise_beacon.go:498]: i think this is the case. i don't see any code that prevents adding duplicate proposals

File: tortoisebeacon/tortoise_beacon.go (hunk 2 above)
Comment (dshulyak) [reply tortoisebeacon/tortoise_beacon.go]: @countvonzero i updated the code. why do you think that we should:
> the late vote received after weak coin started will be added to the next round

why not to ignore that vote? in the next round we will have a next set of votes

**Comment by dshulyak** [comment]
I guess i will need to address the problem of slow discovery to understand if there is any additional flakiness caused by this PR, previous test failed exactly the same way. One node wasn't able to start sync on time and atx in epoch 1 wasn't published.

**Comment by dshulyak** [comment]
bors try

File: tortoisebeacon/tortoise_beacon.go (hunk 2 above)
Comment (countvonzero) [reply tortoisebeacon/tortoise_beacon.go]: that was basedd on Tal in the R&D meeting. my impression is tgat we want as many vote as possible to cross threshold. as long as a node isnt double voting, we should count it. but if that introduce extra code complexity, we should talk to Tal abt this

Comment (dshulyak) [reply tortoisebeacon/tortoise_beacon.go]: i think thats possible in general but we need to revisit how votes are buffered. but what i don't understand is what we should do with the vote that will be sent in the next round in such case? i don't think that we should count votes from the same smesher twice (e.g. first time from late vote in the previous round and second time from timely vote in the current round)

**Comment by dshulyak** [comment]
bors try

File: tortoisebeacon/tortoise_beacon.go (hunk 2 above)
Comment (countvonzero) [reply tortoisebeacon/tortoise_beacon.go]: it's not so much `count votes from the same smesher twice` tho. essentially every smesher gets to vote K times where K is the number of rounds. 

if we consider a honest smesher (not casting multiple votes for the same round), then it's just network delay and there is not reason to _not_ count it.

but if a smesher is malicious, casting late votes (after it learns the weakcoin) can potentially sway the vote to cross the threshold early in that round.

my guess would be it's OK to just drop late votes for code simplicity. but it's worth to double-check with Tal maybe in tuesday R&D meeting

**Comment by dshulyak** [comment]
bors try

**Comment by dshulyak** [comment]
bors try

**Comment by dshulyak** [comment]
bors try

**Comment by dshulyak** [comment]
bors try

**Comment by dshulyak** [comment]
bors try

**Comment by dshulyak** [comment]
bors try

**Comment by dshulyak** [comment]
bors try

**Comment by dshulyak** [comment]
bors try

**Comment by dshulyak** [comment]
bors try

File: tortoisebeacon/tortoise_beacon.go (hunk 2 above)
Comment (dshulyak) [reply tortoisebeacon/tortoise_beacon.go]: sure lets discuss it. i can also open a follow up issue to refactor buffering to make counting late votes easier

**Comment by dshulyak** [comment]
bors try

**Comment by dshulyak** [comment]
bors merge

**Pull request closed without merging**